### DISCUSSION TITLE
Help with Custom Treesitter Capture Groups in mini.ai for JSX Attributes

### DISCUSSION THREAD

**Discussion by Demianeen** [nvim-mini/mini.nvim#623] [Q&A] [answered]
I'm trying to define custom text objects for JSX attributes using mini.ai. The goal is to create text objects for the outer and inner parts of JSX attributes.

I tried lazyvim [textobjects](https://www.lazyvim.org/plugins/coding#miniai) and they are working. In a similar way I tried to define JSX attributes treesitter object:
lua
```
{
  custom_textobjects = {
    A = ai.gen_spec.treesitter({
      a = '@jsx_attribute.outer',
      i = '@jsx_attribute.inner',
    }, {}),
  },
}
```

But I am receiving error message that textobject is not found nearby (even if I am). The specific error message I receive is:
(mini.ai) No textobject "aA" found covering region within 50 lines and 'search_method = 'cover_or_next'.

Could you please advise on correctly defining and using these capture groups for JSX attributes in mini.ai? The issue may be with the way I've defined the capture groups or if there's something I'm missing in the Treesitter setup for JSX.

**Comment by echasnovski** [comment, marked as answer]
LazyVim uses captures which come from 'nvim-treesitter-textobjects' queries. Basically, a curated collections of common filetype textobjects tailored to how certain tree-sitter parsers define their structure (node names, etc.).

Judging by [this file](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/queries/jsx/textobjects.scm) for 'jsx' filetype, I'd try `@parameter.outer` and `@parameter.inner` instead of `@jsx_attribute.outer` and `@jsx_attribute.inner`.

Also make sure that the target textobject indeed can be seen within 50 lines of current line. If it is bigger than that, increase `config.n_lines`.

**Reply by Demianeen** [reply]
But why does the mini.ai argument text object behave differently? I tried to read the argument function, but it is too hard for me 😅

**Reply by echasnovski** [reply]
Built-in 'argument' textobject doesn't depend on tree-sitter to work in cases when tree-sitter parser is not active. Starting from files without tree-sitter enabled and ending with strings and comments.

**Comment by xixiaofinland** [comment]
I also wish to switch from `nvim-treesitter-textobjects` to `mini.ai`. I was also stuck here, right about to check `textobjects.scm`.  IMO, the error "No textobject "aA" found..." is misleading...

Question: does _mini.ai_ also have jump action like in _nvim-treesitter-textobjects_ we can define keys to jump/move to the next matching capture? @echasnovski

**Reply by echasnovski** [reply]
> IMO, the error "No textobject "aA" found..." is misleading...

Maybe slightly, but not completely. As there indeed no textobject can be found when supplied some not appropriate captures.

> Question: does _mini.ai_ also have jump action like in _nvim-treesitter-textobjects_ we can define keys to jump/move to the next matching capture? @echasnovski

Not as detailed, but [yes](https://github.com/echasnovski/mini.nvim/blob/ea1af8c7d5e72148cae8a04e9887322a53fe66cf/doc/mini-ai.txt#L416-L418). Mostly designed for moving to left/right edge when inside the textobject.

**Comment by xixiaofinland** [comment]
> Not as detailed, but [yes](https://github.com/echasnovski/mini.nvim/blob/ea1af8c7d5e72148cae8a04e9887322a53fe66cf/doc/mini-ai.txt#L416-L418). Mostly designed for moving to left/right edge when inside the textobject.

ahhhh... the `g[`/`g]` is similar to the "move" action.... Finally it ticks! Thank you!

**Comment by xixiaofinland** [comment]
one more related question, the `textobjects.scm` I use is out-of-box from the nvim-treesitter-textobjects. It seems `nimi.ai` can't use the captures in it automatically, can it? @echasnovski

**Reply by echasnovski** [reply]
Not sure what you mean by "automatically". If captures are "reachable" (as described [here](https://github.com/echasnovski/mini.nvim/blob/ea1af8c7d5e72148cae8a04e9887322a53fe66cf/doc/mini-ai.txt#L689-L698)), it should work.

**Reply by xixiaofinland** [reply]
Thanks, question actually answered by your link.
Happy Xmas!
